File: Chapter-6/queue_imp_using_array.cpp
#include<bits/stdc++.h>

using namespace std;

#define N 5
int Que[N];
int rear = -1;
int front = -1;

void EnQueue(){
    int item;
    if(rear == N-1){
        cout<<" Overflow! "<<endl;
    }
    else if(rear == -1 && front ==-1){
        cout<<"enter the data: ";
        cin>>item;
        rear = front = 0;
        Que[rear] = item;
    }
    else{
        cout<<"enter the data: ";
        cin>>item;
        rear++;
        Que[rear] = item;
    }
}

void DeQueue(){
    if(rear==-1 && front==-1){
        cout<<"Underflow!"<<endl;
    }
    else if(rear==front){
        rear = front = -1;
    }
    else{
        cout<<"Dequeue element is: "<<Que[front]<<endl;
        front++;
    }
}

void Peek(){
    if(rear==-1 && front==-1){
        cout<<"Underflow!"<<endl;
    }
    cout<<"Peek element is: "<<Que[front]<<endl;
}


void Display(){
    int temp = front;
    if(rear==-1 && front==-1){
        cout<<"Underflow!"<<endl;
    }
    else{
        while(temp<=rear){
            cout<<Que[temp]<<" ";
            temp++;
        }
        cout<<endl;
    }
}



int main(){
    int task;
    do{
        cout<<"1. Enqueue in queue"<<endl;
        cout<<"2. dequeue in a queue"<<endl;
        cout<<"3. peak in queue"<<endl;
        cout<<"4. Display The queue"<<endl;
        cout<<"0. For Exit"<<endl;
        cout<<" enter an option : ";
        cin>>task;
        switch(task){
        case 0:
            break;
        case 1:
            EnQueue();
            cout<<endl;
            break;

        case 2:
            DeQueue();
            cout<<endl;
            break;

        case 3:
            Peek();
            cout<<endl;
            break;

        case 4:
            Display();
            cout<<endl;
            break;

        default:
            cout<<"please enter a correct choice"<<endl;
            cout<<endl;
            break;
        }
    }while(task!=0);

    return 0;
}

File: Chapter-6/queue_imp_using_array.py
class Que:
    element = []
    front = rear = -1

    def EnQueue(self, item):
        if self.rear == -1 and self.front == -1:
            self.rear = self.front = 0
            self.element.append(item)
        else:
            self.rear += 1
            self.element.append(item)

    def DeQueue(self):
        if self.front == -1 and self.rear == -1:
            print(" Underflow! ")
        else:
            print(f"Dequeue element : {self.element[self.front]}")
            self.front += 1

    def Peek(self):
        if self.front == -1 and self.rear == -1:
            print(" Underflow! ")
        else:
            print(f"Peek element is: {self.element[self.front]}")

    def Display(self):
        temp = self.front
        if self.front == -1 and self.rear == -1:
            print(" Underflow! ")
        else:
            while temp <= self.rear:
                print(self.element[temp], end=" ")
                temp = temp + 1
            print()


Q = Que()
print("Elements in the queue are :")
Q.EnQueue(89)
Q.EnQueue(23)
Q.EnQueue(35)
Q.EnQueue(76)
Q.EnQueue(10)
Q.Display()

Q.DeQueue()
Q.Display()
Q.Peek()
